driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java

```diff
@@ -388,20 +389,44 @@ private Connection findLeastBusyForShard(int shardId) {
   }
 
   ListenableFuture<Connection> borrowConnection(
-      long timeout, TimeUnit unit, int maxQueueSize, ByteBuffer routingKey) {
+      long timeout, TimeUnit unit, int maxQueueSize, Statement statement) {
+
     Phase phase = this.phase.get();
     if (phase != Phase.READY)
       return Futures.immediateFailedFuture(
           new ConnectionException(host.getSocketAddress(), "Pool is " + phase));
 
     int shardId = 0;
-    if (host.getShardingInfo() != null) {
+    if ((host.getShardingInfo() != null) && (statement != null)) {
+      Metadata metadata = manager.cluster.getMetadata();
+      ProtocolVersion protocolVersion =
+          manager.cluster.getConfiguration().getProtocolOptions().getProtocolVersion();
+      CodecRegistry codecRegistry = manager.cluster.getConfiguration().getCodecRegistry();
+      ByteBuffer routingKey = statement.getRoutingKey(protocolVersion, codecRegistry);
       if (routingKey != null) {
-        Metadata metadata = manager.cluster.getMetadata();
         Token t = metadata.newToken(routingKey);
         shardId = host.getShardingInfo().shardId(t);
       } else {
-        shardId = RAND.nextInt(host.getShardingInfo().getShardsCount());
+        String csTokens = TokenAwarePolicy.getTokenRange(statement);
+        if (csTokens != null) {
+          String[] strTokens = csTokens.split(",");
+          Token start = metadata.newToken(strTokens[0]);
+          Token end = metadata.newToken(strTokens[1]);
+          int startShardId = host.getShardingInfo().shardId(start);
+          int endShardId = host.getShardingInfo().shardId(end);
+          if (startShardId == endShardId) {
+            shardId = startShardId;
+          } else {
+            if (endShardId < startShardId) {
```

Comment:
I think it might make sense to always select startShardId because that's where the request will end up I believe. That's something to be confirmed though.

Reply:
Here's my reasoning:

1) if shardId(startToken) == shardId(endToken), then we assume the token range lies entirely in a shard. Hence, we rely on shardId(startToken).

2) if shardId(startToken) != shardId(endToken), then we know the token range extrapolates a single shard. In this case, I pick a random shardId between shardId(startToken) and shardId(endToken). I do this because if the client is mistakenly using too large ranges, we at least try to randomly select a shard as a non shard-aware driver would do.

Deos that make sense?